src/track/beats.cpp

```diff
@@ -32,25 +541,474 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+    FramePos upper_bound = findNthBeat(lower_bound, n * 2).getFramePosition();
+    if (upper_bound == kInvalidFramePos) {
+        upper_bound = m_beats.last().getFramePosition();
+        lower_bound = findNthBeat(upper_bound, n * -2).getFramePosition();
+        // Super edge-case -- the track doesn't have n beats!  Do the best
+        // we can.
+        if (lower_bound == kInvalidFramePos) {
+            lower_bound = m_beats.first().getFramePosition();
+        }
+    }
+
+    // TODO(JVC) We are extracting frame numbers to then construct beats.
+    // Then in calculateBpm we are using the frame position to find
+    // the beats to  use them to calculate. Seems inefficient
+    // Will not make more sense to extract the Beats straight?
+    // We can use getBpmRange and move the logic of calculateBpm there
+    Beat startBeat(lower_bound), stopBeat(upper_bound);
+    return calculateBpm(startBeat, stopBeat);
+}
+
+void BeatsInternal::setSignature(TimeSignature sig, int downbeatIndex) {
+    if (!isValid()) {
+        return;
+    }
+    track::io::TimeSignatureMarker markerToInsert;
+    markerToInsert.set_downbeat_index(downbeatIndex);
+    markerToInsert.mutable_time_signature()->set_beats_per_bar(
+            sig.getBeatsPerBar());
+    markerToInsert.mutable_time_signature()->set_note_value(sig.getNoteValue());
+    QVector<track::io::TimeSignatureMarker> timeSignatureMarkersMutableCopy;
+    copy(m_beatsProto.time_signature_markers().cbegin(),
+            m_beatsProto.time_signature_markers().cend(),
+            std::back_inserter(timeSignatureMarkersMutableCopy));
+
+    track::io::TimeSignatureMarker searchBeforeMarker;
+    searchBeforeMarker.set_downbeat_index(downbeatIndex);
+    auto prevTimeSignatureMarker =
+            std::lower_bound(timeSignatureMarkersMutableCopy.begin(),
+                    timeSignatureMarkersMutableCopy.end(),
+                    searchBeforeMarker,
+                    TimeSignatureMarkerEarlier);
+    if (prevTimeSignatureMarker->downbeat_index() == downbeatIndex) {
+        prevTimeSignatureMarker->CopyFrom(markerToInsert);
+    } else {
+        timeSignatureMarkersMutableCopy.insert(
+                prevTimeSignatureMarker, markerToInsert);
+    }
+    m_beatsProto.clear_time_signature_markers();
+    for (const auto& timeSignatureMarker : timeSignatureMarkersMutableCopy) {
+        m_beatsProto.add_time_signature_markers()->CopyFrom(
+                timeSignatureMarker);
+    }
+    generateBeatsFromMarkers();
+}
+
+void BeatsInternal::translate(FrameDiff_t numFrames) {
+    if (!isValid()) {
+        return;
+    }
+
+    m_beatsProto.set_first_frame_position(
+            m_beatsProto.first_frame_position() + numFrames);
+    generateBeatsFromMarkers();
+}
+
+void BeatsInternal::setBpm(Bpm bpm, int beatIndex) {
+    // TODO(hacksdump): A check for preferences will be added to only allow setting bpm
+    //  when "Assume Constant Tempo" is checked.
```

Comment:
This option should be removed, so Beats should not have anything to do with this option.